### PR TITLE
UCS/MEMORY: Added VFS nodes representing UCS rcache.

### PR DESCRIPTION
**Pull request by rakhmets**
## What
Added directory to VFS tree representing UCS rcache object.
Added number of regions of UCS rcache to VFS structure.

## Why ?
The changes help to perform runtime analysis of UCX-based applications.

### CODE REVIEW

File: src/ucs/memory/rcache.c [hunk 1]
```diff
@@ -999,6 +1000,14 @@ static void ucs_rcache_global_list_remove(ucs_rcache_t *rcache) {
     pthread_mutex_unlock(&ucs_rcache_global_list_lock);
 }
 
+static void
+ucs_rcache_vfs_show_num_regions(void *obj, ucs_string_buffer_t *strb)
+{
+    ucs_rcache_t *rcache = obj;
+
+    ucs_string_buffer_appendf(strb, "%lu\n", rcache->num_regions);
+}
```
Comment (yosefe) [inline src/ucs/memory/rcache.c:1009]: 1. IMO we should add generic function ucs_vfs_show_ulong with custom argument (pointer to the ulong value)
2. print also total size, max regions, max size
3. print length of inv_q and gc_list

Comment (rakhmets) [reply src/ucs/memory/rcache.c:1009]: The function is called from vfs. Do you suggest to store the pointer to the value inside the vfs node?

E.g.:
```C
static void
ucs_vfs_read_ro_file(ucs_vfs_node_t *node, ucs_string_buffer_t *strb)
{
    ucs_vfs_node_t *parent_node = node->parent;

    ucs_assert(ucs_vfs_check_node(node, UCS_VFS_NODE_TYPE_RO_FILE) == 1);

    while (ucs_vfs_check_node(parent_node, UCS_VFS_NODE_TYPE_SUBDIR)) {
        parent_node = parent_node->parent;
    }

    ucs_spin_unlock(&ucs_vfs_obj_context.lock);

- node->text_cb(parent_node->obj, strb);
+ node->text_cb.cb(parent_node->text_cb.value, strb);

    ucs_spin_lock(&ucs_vfs_obj_context.lock);
}
```

Comment (rakhmets) [reply src/ucs/memory/rcache.c:1009]: We can use node::obj to store the pointer to the value, but this doesn't look obvious, IMO.

Comment (yosefe) [reply src/ucs/memory/rcache.c:1009]: IMO, like this:
```c
node->text_cb(parent_node->obj, node->arg, strb);
```
then:
```c
static void
ucs_vfs_show_ulong(void *obj, void *arg, ucs_string_buffer_t *strb)
{
   ucs_string_buffer_appendf(strb, "%lu\n", *(unsigned long*)arg);
}

ucs_vfs_obj_add_ro_file(self, ucs_vfs_show_ulong, &rcache->num_regions, "num_regions");
ucs_vfs_obj_add_ro_file(self, ucs_vfs_show_ulong, &rcache->total_size, "total_size");

```

Comment (rakhmets) [reply src/ucs/memory/rcache.c:1009]: got it!

**Comment by rakhmets** [comment]
@yosefe @brminich could you please to review the changes.

**Review by brminich — APPROVED** [review]

File: src/ucs/memory/rcache.c [hunk 2]
```diff
@@ -999,6 +1000,39 @@ static void ucs_rcache_global_list_remove(ucs_rcache_t *rcache) {
...
+    ucs_vfs_obj_add_ro_file(rcache, ucs_vfs_show_memunits, &rcache->total_size,
+                            "total_size");
+    ucs_vfs_obj_add_ro_file(rcache, ucs_vfs_show_ulunits,
+                            &rcache->params.max_regions, "max_regions");
+    ucs_vfs_obj_add_ro_file(rcache, ucs_vfs_show_memunits,
+                            &rcache->params.max_size, "max_size");
+    ucs_vfs_obj_add_ro_file(rcache, ucs_rcache_vfs_show_inv_q_length, NULL,
+                            "inv_q_length");
+    ucs_vfs_obj_add_ro_file(rcache, ucs_rcache_vfs_show_gc_list_length, NULL,
+                            "gc_list_length");
```
Comment (yosefe) [inline src/ucs/memory/rcache.c:1033]: maybe gc_list/length  and inv_q/length?

**Review by yosefe — COMMENTED** [review]
pls squash after fix

**Comment by rakhmets** [comment]
just an example of the ucs sub-tree:
image

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
> just an example of the ucs sub-tree:
> image

BTW, why in knem case the name is "knem/rcache/device"?

**Comment by rakhmets** [comment]
> BTW, why in knem case the name is "knem/rcache/device"?

This was an existing code. All other rcache names use underscores instead of spaces.
The small fix is here #6731.

**Comment by yosefe** [comment]
> > BTW, why in knem case the name is "knem/rcache/device"?
> 
> This was an existing code. All other rcache names use underscores instead of spaces.
> The small fix is here #6731.

Cool! 👍

**Comment by rakhmets** [comment]
bot:pipe:retest

**Pull request merged**